### PR TITLE
Comments API.

### PR DESCRIPTION
**Pull request by dkorolev**
@mzhurovich : I'm not that proud of this code, but has two tests, unit-, and the regression one, run against production.

Thanks!
Dima

### CODE REVIEW

File: server.h [hunk 1]
```diff
@@ -420,15 +443,23 @@ class CTFOServer final {
...
         } else {
           const std::string requested_url = r.url.ComposeURL();
+          const CID cid = RandomCID();
           try {
-            const auto request = ParseJSON<AddCardRequest>(r.body);
+            AddCardRequest request;
+            try {
+              ParseJSON(r.body, request);
+            } catch (const bricks::ParseJSONException&) {
+              const auto short_request = ParseJSON<AddCardShortRequest>(r.body);
```
Comment (mzhurovich) [inline server.h:457]: Maybe add `try/catch` - what if body contains invalid JSON?

File: server.h [hunk 2]
```diff
@@ -487,14 +520,191 @@ class CTFOServer final {
...
+              std::move(r));
+        } else if (r.method == "POST") {
+          const std::string requested_url = r.url.ComposeURL();
+          const OID oid = RandomOID();
+          try {
+            AddCommentRequest request;
+            try {
+              ParseJSON(r.body, request);
+            } catch (const bricks::ParseJSONException&) {
+              const auto short_request = ParseJSON<AddCommentShortRequest>(r.body);
```
Comment (mzhurovich) [inline server.h]: And here?

File: server.h [hunk 3]
```diff
@@ -487,14 +520,191 @@ class CTFOServer final {
...
+    // TODO(dkorolev): Avoid this shameless copy-pasting.
+    const auto comments_handler = [this](Request r) {
+      const UID uid = StringToUID(r.url.query["uid"]);
+      const std::string token = r.url.query["token"];
+      const CID cid = StringToCID(r.url.query["cid"]);
+      if (uid == UID::INVALID_USER) {
+        DebugPrint(Printf("[/ctfo/comment] Wrong UID. Requested URL = '%s'", r.url.ComposeURL().c_str()));
+        r("NEED VALID UID-TOKEN PAIR\n", HTTPResponseCode.BadRequest);
+      } else if (cid == CID::INVALID_CARD) {
+        DebugPrint(Printf("[/ctfo/comments] Wrong CID. Requested URL = '%s'", r.url.ComposeURL().c_str()));
```
Comment (mzhurovich) [inline server.h:533]: I see you created two endpoints for this handler - `/comment` & `/comments` - I suggest to leave one of them in `DebugPrint`'s.

File: server.h (hunk 1 above)
Comment (mzhurovich) [inline server.h:457]: Ah, I see. Ignore the comment :)

File: util.h [hunk 4]
```diff
@@ -62,7 +64,18 @@ inline CID StringToCID(const std::string& s) {
...
-  return CID::INVALID;
+  return CID::INVALID_CARD;
+}
+
+inline std::string OIDToString(const OID oid) {
+  return bricks::strings::Printf("o%020llu", static_cast<uint64_t>(oid));
+}
+
+inline OID StringToOID(const std::string& s) {
+  if (s.length() == 21 && s[0] == 'o') {  // 'c' + 20 digits of `uint64_t` decimal representation.
```
Comment (mzhurovich) [inline util.h]: s/'c'/'o'/

Comment (dkorolev) [inline util.h]: :+1: 4114ce9

File: server.h (hunk 1 above)
Comment (dkorolev) [inline server.h:457]: :-)

File: server.h (hunk 3 above)
Comment (dkorolev) [inline server.h:533]: :+1: 4114ce9

Left `[/ctfo/comments/]` everywhere. The other one is just to be sure.

**Comment by dkorolev** [comment]
@mzhurovich : Thanks for the review -- PTAL!

**Pull request merged**